// This describes the Interface for Export / Import of Rosette object code. This is used
// between Rosette and Roscala for interchanging compiled object code.


syntax = "proto3";

package ObjectCodePB;



message ObjectCode {
    // There are two parts to an object code file.
    //
    // Part one (CodeBlocks) is a repeated list of code export blocks.
    // Part two (Objects) is a repeated list of Rosette objects referenced by the code blocks.

    CodeBlocks codeblocks = 1;
    Objects objects = 2;
}

message CodeBlocks {
    repeated CodeBlock codeblock = 1;
}

// This central objects list is for objects that are immutable, and should be accessed by
// reference.
message Objects {
    repeated Object object = 1;
}

// A single code block contains a litvec with objects on which the code operates,
// and the codevec with the actual opCodes that are executed by the VM.
message CodeBlock {
    LitVec litvec = 1;
    CodeVec codevec = 2;
}

// This message contains the actual executable opCodes
message CodeVec {
    // Opcodes are serialized in binary
    bytes opcodes = 1;
}

message LitVec {
    // Items within a Litvec are accessed by index and must be stored in the same order as
    // they originally existed at compile time.

    repeated LitVecItem item = 1; // This ties to object id's within objects in the ObjectVector
}

// A LitVecItem can either contain an instance of the object itself, or an id that ties it
// to an object in the central objects list.
message LitVecItem {
    oneof item {
        Object ob = 2;
        uint64 id = 3;
    }
}

// Rather than implement all 100+ object types existing in Rosette, We will only implement those
// used in loading boot.rbl, plus any deemed necessary by the Roscala compiler guys.

// TODO: These messages need to be filled out with the required fields. Also if there are objects that
// contain other objects, they will need to be added to this list.

message Actor {}
message BlockExpr {}
message Char {
    uint32 value = 1;
}
message Code {
}
message ExpandedLocation {}
message Fixnum {
    int32 value = 1;
}
message Float {
    double value = 1;
}
message FreeExpr {}
message GotoExpr {}
message IfExpr {}
message LabelExpr {}
message LetExpr {}
message LetrecExpr {}
message MethodExpr {}
message Proc {}
message ProcExpr {}
message QuoteExpr {}
message RBLstring {
    string value = 1;
}
message ReflectiveMethodExpr {}
message RequestExpr {}
message SendExpr {}
message SeqExpr {}
message SetExpr {}
message StdMthd {}
message Symbol {
    string name = 1;
}
message TblObject {}
message Template {}
message Tuple {}
message TupleExpr {}

enum ObType {
    OT_Actor = 0;
    OT_BlockExpr = 1;
    OT_Char = 2;
    OT_Code = 3;
    OT_ExpandedLocation = 4;
    OT_Fixnum = 5;
    OT_Float = 6;
    OT_FreeExpr = 7;
    OT_GotoExpr = 8;
    OT_IfExpr = 9;
    OT_LabelExpr = 10;
    OT_LetExpr = 11;
    OT_LetrecExpr = 12;
    OT_MethodExpr = 13;
    OT_Proc = 14;
    OT_ProcExpr = 15;
    OT_QuoteExpr = 16;
    OT_RBLstring = 17;
    OT_ReflectiveMethodExpr = 18;
    OT_RequestExpr = 19;
    OT_SendExpr = 20;
    OT_SeqExpr = 21;
    OT_SetExpr = 22;
    OT_StdMthd = 23;
    OT_Symbol = 24;
    OT_TblObject = 25;
    OT_Template = 26;
    OT_Tuple = 27;
    OT_TupleExpr = 28;
}

message Object {
    ObType type = 1;
    uint64 id = 2;

    // One and only one of these should be populated
    oneof object {
        Actor actor = 10;
        BlockExpr blockexpr = 11;
        Char char = 12;
        Code code = 13;
        ExpandedLocation expandedlocation = 14;
        Fixnum fixnum = 15;
        Float float = 16;
        FreeExpr freeexpr = 17;
        GotoExpr gotoexpr = 18;
        IfExpr ifexpr = 19;
        LabelExpr labelexpr = 20;
        LetExpr letexpr = 21;
        LetrecExpr letrecexpr = 22;
        MethodExpr methodexpr = 23;
        Proc proc = 24;
        ProcExpr procexpr = 25;
        QuoteExpr quoteexpr = 26;
        RBLstring rblstring = 27;
        ReflectiveMethodExpr reflectivemethodexpr = 28;
        RequestExpr requestexpr = 29;
        SendExpr sendexpr = 30;
        SeqExpr seqexpr = 31;
        SetExpr setexpr = 32;
        StdMthd stdmthd = 33;
        Symbol symbol = 34;
        TblObject tblobject = 35;
        Template template = 36;
        Tuple tuple = 37;
        TupleExpr tupleexpr = 38;
    }
}
